### PR TITLE
Improve application resource caching

### PR DESCRIPTION
**Pull request by muralov**
**Description**

Changes proposed in this pull request:

- A controller is developed in App Connectivity Validator which updates the cache triggered by events of Application CRs. In addition, the controller periodically updates the cache.
- App Connectivity Validator unit tests are updated 
- Application Operator changed in order to give the Cache Sync Controller `watch` right on Application CRs

**Related issue(s)**
#8036

### CODE REVIEW

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (Maladie) [inline components/application-connectivity-validator/internal/controller/cachesynccontroller.go:40]: At the moment controller caches all Applications CRs. While Validator is deployed by Application Operator per single Application CR, we don't really need to cache/watch all Applications.

**Review by Maladie — CHANGES_REQUESTED** [review]

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (muralov) [reply components/application-connectivity-validator/internal/controller/cachesynccontroller.go:40]: Yes, you are right. I am already aware of this and I already mentioned this in the description of the PR as a future improvement. I would like to merge this PR without this enhancement if it is okay? As a next step I will do this improvement by filtering events.

File: components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go
Comment (muralov) [reply components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go:46]: 120 is actually in seconds (i.e. 2 minutes). This is period where controller triggers reconciliation in every 2 minutes in addition to normal event triggers. This means that controller tries to sync every 2 minutes although no events occur on application resources. I thought 2 minutes is event shorter :).

Comment (muralov) [reply components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go:46]: 120 is actually in seconds (i.e. 2 minutes). This is period where controller triggers reconciliation in every 2 minutes in addition to normal event triggers. This means that controller tries to sync every 2 minutes although no events occur on application resources. I thought 2 minutes is event shorter :).

Comment (Maladie) [reply components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go:46]: I've checked this, and it is 120ns. When you check the logs, you will also see that at the start of the Validator, there is logged warning about this.

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (Maladie) [reply components/application-connectivity-validator/internal/controller/cachesynccontroller.go:40]: Oh, sorry, I missed that. It would be great if you could also implement this in this PR.

File: components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go
Comment (muralov) [reply components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go:46]: It somehow got unnoticed because I have been always passing it `-syncPeriod=20s` from outside as an argument. I will fix it.

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (muralov) [reply components/application-connectivity-validator/internal/controller/cachesynccontroller.go:40]: done. Please have a look.

File: components/application-connectivity-validator/internal/controller/signals/signal_posix.go
Comment (muralov) [reply components/application-connectivity-validator/internal/controller/signals/signal_posix.go]: I have moved `shutdownSignals` into the go file where it is used. At least we avoid navigating to another go file :)

**Comment by muralov** [comment]
@Maladie any update on the PR? It'd be perfect if we can merge PR in this week because from Saturday until mid January I may not be able to work on the issue.

File: components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go
Comment (koala7659) [inline components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go:48]: The value of syncPeriod is not used in the helm chart in application-operator (file: components/application-operator/charts/application/templates/validator.yaml)

File: components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go
Comment (koala7659) [inline components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go:30]: Please add a description of new parameters to the file components/application-connectivity-validator/README.md

**Review by koala7659 — COMMENTED** [review]
Sorry for keeping you wait so long. I'm providing some comments now but we are still discussing your solution among the team.

Please make sure to remove all comments from the code before merging

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (koala7659) [inline components/application-connectivity-validator/internal/controller/cachesynccontroller.go:80]: When application is deleted, the application-connectivity-validator as its part will we deleted as well. Maybe we don't need to react to this event since the pod itself will be deleted.

File: components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go
Comment (koala7659) [reply components/application-connectivity-validator/cmd/applicationconnectivityvalidator/options.go:48]: But maybe this is OK

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (koala7659) [reply components/application-connectivity-validator/internal/controller/cachesynccontroller.go]: Fixed

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (koala7659) [reply components/application-connectivity-validator/internal/controller/cachesynccontroller.go:141]: Fixed

File: components/application-connectivity-validator/internal/controller/cachesynccontroller.go
Comment (koala7659) [reply components/application-connectivity-validator/internal/controller/cachesynccontroller.go:106]: I have decided not to change the code, since it is taken from idiomatic template. I have just removed comments

**Review by koala7659 — DISMISSED** [review]
I believe the implementation is correct. I'm sorry that this review procedure has been so long.

**Review by Maladie — DISMISSED** [review]

**Review by franpog859 — APPROVED** [review]

File: components/application-connectivity-validator/README.md
Comment (kazydek) [inline components/application-connectivity-validator/README.md:19]:
```suggestion
- **kubeConfig** is the path to a cluster kubeconfig. Used for running the service outside of the cluster.
```

File: components/application-connectivity-validator/README.md
Comment (kazydek) [inline components/application-connectivity-validator/README.md:20]:
```suggestion
- **masterURL** is the address of the Kubernetes API server. Overrides any value in kubeconfig. Used for running the service outside of the cluster.
```

**Review by kazydek — APPROVED** [review]
🎉

**Review by akgalwas — APPROVED** [review]

**Pull request merged**
